File: 백준/Gold/2749. 피보나치 수 3/피보나치 수 3.py
def fibonacci_modulo(n, m):
    if n <= 1:
        return n % m

    previous, current = 0, 1
    for _ in range(n - 1):
        previous, current = current, (previous + current) % m

    return current


def pisano_period(m):
    previous, current = 0, 1
    for i in range(m * 6):
        previous, current = current, (previous + current) % m
        if previous == 0 and current == 1:
            return i + 1


n = int(input())
mod_value = 1000000

period_length = pisano_period(mod_value)
result = fibonacci_modulo(n % period_length, mod_value)

print(result)

File: 백준/Gold/2749. 피보나치 수 3/피보나치 수 3.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        long modValue = 1000000;

        long periodLength = pisanoPeriod(modValue);
        long result = fibonacciModulo(n % periodLength, modValue);

        System.out.println(result);
    }

    static long fibonacciModulo(long n, long m) {
        if (n <= 1) {
            return n % m;
        }

        long previous = 0, current = 1;
        for (long i = 0; i < n - 1; i++) {
            long temp = current;
            current = (previous + current) % m;
            previous = temp;
        }

        return current;
    }

    static long pisanoPeriod(long m) {
        long previous = 0, current = 1;
        for (long i = 0; i < m * 6; i++) {
            long temp = current;
            current = (previous + current) % m;
            previous = temp;

            if (previous == 0 && current == 1) {
                return i + 1;
            }
        }

        return -1;
    }
}